lib/handshake_client.ml

```diff
@@ -141,9 +141,9 @@ let answer_server_hello state (ch : client_hello) sh secrets raw log =
 
   let* () =
     if max_protocol_version state.config.protocol_versions = `TLS_1_3 then
-      let piece = Cstruct.sub sh.server_random 24 8 in
-      let* () = guard (not (Cstruct.equal Packet.downgrade12 piece)) (`Fatal `Downgrade12) in
-      guard (not (Cstruct.equal Packet.downgrade11 piece)) (`Fatal `Downgrade11)
+      let piece = String.sub sh.server_random 24 8 in
```

Comment:
allocates, maybe we have a function in eqaf or define our own to have off and len params for eq?

Reply:
I did a try about implementations of `String.sub_equal` (which does not guarantee the constant time as `eqaf`). My benchmark is available here: https://gist.github.com/dinosaure/7ef54e9431217435e3b81df890192b72. From my results (on AMD), the "naive" implementation (the Oracle one) is fast enough for the worst case. For readability, I will consider this implementation for `ocaml-tls` (into the `lib/utils.ml` file) than the other.

Reply:
Actually, and according to my benchmark, it seems that an `equal_sub` which uses `String.sub` or an `equal_sub` which tries to iterate into given `strings` have the same performance. My conclusion is: even if we do a `String.sub` here, try to avoid this `String.sub` and try to directly read bytes and compare them is not faster than the `String.sub`...